### ISSUE TITLE
Assertation in txn_limbo_read_promote fails.

### ISSUE THREAD

**Issue by grafin** [tarantool/tarantool#6842]
src/box/txn_limbo.c:515: txn_limbo_read_promote: 365 Assertion `txn_limbo_is_empty(&txn_limbo)' failed.

- OS: Linux
- OS Version: nixos 21.11
- Architecture: amd64

`tarantool --version` output:

```
Tarantool 2.10.0-beta2-90-g18eae244c
Target: Linux-x86_64-Debug
Build options: cmake . -DCMAKE_INSTALL_PREFIX=/home/boris/dev/github/tarantool/build/install -DENABLE_BACKTRACE=ON
Compiler: /nix/store/n5vkys6wvrfnkd381d39y5926rpqpk4w-gcc-wrapper-10.3.0/bin/gcc /nix/store/n5vkys6wvrfnkd381d39y5926rpqpk4w-gcc-wrapper-10.3.0/bin/g++
C_FLAGS: -fexceptions -funwind-tables -fno-omit-frame-pointer -fno-stack-protector -fno-common -fopenmp -msse2 -fprofile-arcs -ftest-coverage -std=c11 -Wall -Wextra -Wno-strict-aliasing -Wno-char-subscripts -Wno-format-truncation -Wno-gnu-alignof-expression -fno-gnu89-inline -Wno-cast-function-type -Werror
CXX_FLAGS: -fexceptions -funwind-tables -fno-omit-frame-pointer -fno-stack-protector -fno-common -fopenmp -msse2 -fprofile-arcs -ftest-coverage -std=c++11 -Wall -Wextra -Wno-strict-aliasing -Wno-char-subscripts -Wno-format-truncation -Wno-invalid-offsetof -Wno-gnu-alignof-expression -Wno-cast-function-type -Werror
```

**Steps to reproduce**

Run following test MANY times (fails 1-2 times on 1000 runs).

<details><summary>bug_test.lua</summary>

```lua
local luatest = require('luatest')
local helpers = require('test.luatest_helpers')
local cluster = require('test.luatest_helpers.cluster')
local g = luatest.group('bug')
local cluster_init = function(g)
    g.cluster = cluster:new({})

    g.box_cfg = {
        election_mode = 'off',
        read_only = false,
        replication_timeout = 0.1,
        replication_synchro_timeout = 5,
        replication_synchro_quorum = 1,
        replication = {
            helpers.instance_uri('server_', 1),
            helpers.instance_uri('server_', 2),
        },
    }

    g.server_1 = g.cluster:build_and_add_server(
        {alias = 'server_1', box_cfg = g.box_cfg})
    g.server_2 = g.cluster:build_and_add_server(
        {alias = 'server_2', box_cfg = g.box_cfg})
    g.cluster:start()
end

g.before_all(cluster_init)
g.after_all(function(g)
    g.cluster:drop()
end)

-- Demoting should fail if it is interrupted from another server
-- while waiting for synchro queue being emptied.
g.test_limbo_full_interfering_demote = function(g)
    g.server_2:exec(function()
        box.ctl.promote()
    end)

    g.server_2:exec(function()
        box.schema.create_space('test', {is_sync = true}):create_index('pk')
        box.cfg{
            replication_synchro_quorum = 3,
            replication_synchro_timeout = 1000,
        }
    end)

    g.server_1:exec(function()
        box.cfg{
            replication_synchro_timeout = 0.1,
        }
    end)

    g.server_2:exec(function()
        local s = box.space.test
        require('fiber').create(s.replace, s, {1}):id()
    end)

    -- Start demoting server_2 and interrupt it from server_1
    local f = g.server_2:exec(function()
        local f = require('fiber').new(function() box.ctl.demote() end)
        f:set_joinable(true)
        return f:id()
    end)
    g.server_1:exec(function()
        pcall(box.ctl.promote)
    end)
    require('fiber').sleep(1)
    g.server_2:exec(function(f)
        require('fiber').find(f):join()
    end, {f})
end
```

</details>

How to run:
./test/test-run.py --builddir=$(realpath ../) $(yes replication-luatest/bug | head -n800) --no-output-timeout -1 --replication-sync-timeout -1 --test-timeout -1

**Actual behavior**
Assertion fails.
test/var/rejects/replication-luatest/gh_6033_box_promote_demote.reject:
```
TAP version 13
1..1
# Started on Mon Feb  7 23:31:40 2022
# Starting group: bug
server_1 | tarantool: /home/boris/dev/github/tarantool/src/box/box.cc:1781: void box_issue_promote(uint32_t, int64_t): Assertion `raft->volatile_term == raft->term' failed.
not ok 1        bug.test_limbo_full_interfering_demote
#   Peer closed
#   stack traceback:
#       builtin/box/net_box.lua:727: in function '_request'
#       builtin/box/net_box.lua:779: in function 'eval'
#       .../replication-luatest/bug_test.lua:64: in function 'bug.test_limbo_full_interfering_demote'
#       ...
#       [C]: in function 'xpcall'
# Ran 1 tests in 2.587 seconds, 0 successes, 1 error
```

**Expected behavior**
Test always passes.

**Comment by Gerold103** [comment]
This test crashes much more often:
<details><summary>test</summary>

```Lua
local luatest = require('luatest')
local server = require('test.luatest_helpers.server')
local cluster = require('test.luatest_helpers.cluster')
local g = luatest.group('bug')
local cluster_init = function(g)
    g.cluster = cluster:new({})

    g.box_cfg = {
        election_mode = 'off',
        read_only = false,
        replication_timeout = 0.1,
        replication_synchro_timeout = 5,
        replication_synchro_quorum = 1,
        replication = {
            server.build_instance_uri('server_1'),
            server.build_instance_uri('server_2'),
        },
    }

    g.server_1 = g.cluster:build_and_add_server(
        {alias = 'server_1', box_cfg = g.box_cfg})
    g.server_2 = g.cluster:build_and_add_server(
        {alias = 'server_2', box_cfg = g.box_cfg})
    g.cluster:start()
end

g.before_all(cluster_init)
g.after_all(function(g)
    g.cluster:drop()
end)

-- Demoting should fail if it is interrupted from another server
-- while waiting for synchro queue being emptied.
g.test_limbo_full_interfering_demote = function(g)
    g.server_2:exec(function()
        box.ctl.promote()
    end)

    g.server_2:exec(function()
        box.schema.create_space('test', {is_sync = true}):create_index('pk')
        box.cfg{
            replication_synchro_quorum = 3,
            replication_synchro_timeout = 1000,
        }
    end)

    g.server_1:exec(function()
        box.cfg{
            replication_synchro_timeout = 0.1,
        }
    end)

    g.server_2:exec(function()
        box.error.injection.set("ERRINJ_RELAY_FASTER_THAN_TX", true)
        local s = box.space.test
        require('fiber').create(s.replace, s, {1}):id()
    end)
    require('fiber').sleep(1)

    -- Start demoting server_2 and interrupt it from server_1
    local f = g.server_2:exec(function()
        local f = require('fiber').new(function() box.ctl.demote() end)
        f:set_joinable(true)
        return f:id()
    end)
    g.server_1:exec(function()
        pcall(box.ctl.promote)
    end)
    require('fiber').sleep(1)
    g.server_2:exec(function(f)
        box.error.injection.set("ERRINJ_RELAY_FASTER_THAN_TX", false)
        require('fiber').find(f):join()
    end, {f})
end
```
</details>
The crash is in server2 when the error injection is removed. I am still digging the reason. But can't understand why it happens yet.

**Comment by Gerold103** [comment]
100% reproducer. Although I think it is a different bug leading to the same assert:
```Lua
--
-- Instance 1
--
-- Step 1
--
box.cfg{
    listen = 3313,
    replication = {3313, 3314},
}
box.schema.user.grant('guest', 'super')

-- Step 3
--
box.ctl.promote()
box.cfg{
    replication_synchro_quorum = 3,
    replication_synchro_timeout = 1000,
}
s = box.schema.create_space('test', {is_sync = true})
_ = s:create_index('pk')
box.error.injection.set('ERRINJ_WAL_DELAY_COUNTDOWN', 0)

-- Step 5
--
fiber = require('fiber')
f1 = fiber.create(function() s:replace{1} end)
while not box.error.injection.get('ERRINJ_WAL_DELAY') do fiber.sleep(0.1) end
box.error.injection.set('ERRINJ_WAL_DELAY', false)


--
-- Instance 1
--
-- Step 2
--
box.cfg{
    listen = 3314,
    replication = {3313, 3314},
    replication_synchro_timeout = 1000,
    read_only = true,
}

-- Step 4
--
box.ctl.promote()
```
The problem is that when promote is written to WAL and wants to do stuff in the limbo, it didn't expect something was added to the limbo during the WAL write.

**Comment by Gerold103** [comment]
Probably we need a stronger thing than just a promote-latch on the limbo. It should be a read-write lock with preference to writers. "Readers" are transactions going into the queue. "Writers" are limbo state changers - promote coming from an applier, local promote/demote. "Readers" take the read-lock until WAL write is done, not until commit. If there is a pending "writer", new "readers" are not accepted.

That way when PROMOTE/DEMOTE are coming, they will always operate on a fully stable limbo without rows which are between TX and WAL.

Another issue besides assertion, demonstrated in the test above, is that there was PROMOTE from another node going to WAL and yet the limbo managed to allow putting another transaction into self. After PROMOTE is written to WAL, that transaction becomes invalid - it comes not from the limbo owner and yet is written after foreign PROMOTE in WAL.
